relay-server/src/services/spooler/envelope_stack/sqlite.rs

```diff
@@ -0,0 +1,684 @@
+use crate::envelope::Envelope;
+use crate::extractors::StartTime;
+use crate::services::spooler::envelope_stack::EnvelopeStack;
+use crate::utils::get_sampling_key;
+use futures::StreamExt;
+use relay_base_schema::project::ProjectKey;
+use sqlx::query::Query;
+use sqlx::sqlite::{SqliteArguments, SqliteRow};
+use sqlx::{Pool, QueryBuilder, Row, Sqlite};
+use std::collections::VecDeque;
+use std::error::Error;
+use std::fmt::Debug;
+use std::num::NonZeroUsize;
+use std::pin::pin;
+
+/// An error returned when doing an operation on [`SQLiteEnvelopeStack`].
+#[derive(Debug, thiserror::Error)]
+pub enum SQLiteEnvelopeStackError {
+    /// The stack is empty.
+    #[error("the stack is empty")]
+    Empty,
+
+    /// The database encountered an unexpected error.
+    #[error("a database error occurred")]
+    DatabaseError(#[from] sqlx::Error),
+
+    /// The envelope has the project keys that are not matching the ones of the stack.
+    #[error("the envelope doesn't have matching project keys with the stack")]
+    MismatchedEnvelope,
+}
+
+/// An [`EnvelopeStack`] that is implemented on an SQLite database.
+///
+/// For efficiency reasons, the implementation has an in-memory buffer that is periodically spooled
+/// to disk in a batched way.
+pub struct SQLiteEnvelopeStack {
+    /// Shared SQLite database pool which will be used to read and write from disk.
+    db: Pool<Sqlite>,
+    /// Threshold defining the maximum number of envelopes in the `batches_buffer` before spooling
+    /// to disk will take place.
+    spool_threshold: NonZeroUsize,
+    /// Size of a batch of envelopes that is written to disk.
+    batch_size: NonZeroUsize,
+    /// The project key of the project to which all the envelopes belong.
+    own_key: ProjectKey,
+    /// The project key of the root project of the trace to which all the envelopes belong.
+    sampling_key: ProjectKey,
+    /// In-memory stack containing all the batches of envelopes that are read and written to disk.
+    #[allow(clippy::vec_box)]
+    batches_buffer: VecDeque<Vec<Box<Envelope>>>,
+    /// The total number of envelopes inside the `batches_buffer`.
+    batches_buffer_size: usize,
```

Comment:
We need to iterate over the entire stack and count the lengths of the arrays, it is technically constant since it's based on the size of the stack which never grows because we spool